cnrancher/octopus
# [WIP] feat: add agent-device adaptor

**Fixes:**
https://github.com/cnrancher/octopus/issues/50



**Problem:**
Manage a group of devices (1…x) which are running all the same containerized applications on devices with the same architecture. 


**Solution:**

- Add `agentDeviceGroup` and `agentDevice` CRDs, use `daemonSet` with `nodeSelector` to manage apps

- When an agentDeviceGroup is created, nodes can join the group by running shell command created in `agentdevicegroup.status.command` on the target node. 


- The adaptor creates `AgentDevice` when new nodes are labeled with `devices.edge.cattle.io/group` = `dl.metadata.name`, and run daemonSets on these nodes for applications specified in `template.spec.apps`.  



**Test plan:**

1. create `devicelink` for `agentdevicegroup`. e.g.
```yaml
apiVersion: edge.cattle.io/v1alpha1
kind: DeviceLink
metadata:
  name: ii
spec:
  adaptor:
    node: master
    name: adaptors.edge.cattle.io/agent
  model:
    apiVersion: "devices.edge.cattle.io/v1alpha1"
    kind: "AgentDeviceGroup"
  template:
    metadata:
      labels:
        device: ip-172-31-31-2
    spec:
      apps:
        - name: nginx
          namespace: default
          template:
            metadata:
              labels:
                device: ip-172-31-31-2
            spec:
              containers:
                - name: nginx
                  image: nginx:1.14.2
                  ports:
                    - containerPort: 80
```

2. check created `agentdevicegroup`, run `status.command` on the target node

1.  wait for node up and check the node label to see `devices.edge.cattle.io/group` set

1.  check `agentdevice` created, and check the apps successfully deployed on the node
